h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -0,0 +1,720 @@
+        for (int k=0;k<nnzCount;++k) {
+            SparseItem item = new SparseItem();
+            int localIdx = nnz[k];
+            item.pos = (int)ck.start() + localIdx;
+            // both 0 and NA are omitted in the sparse DMatrix
+            if (zw.zeroWeight(localWeight ? localIdx : item.pos)) continue;
+            if (ck.isNA(localIdx)) continue;
+            item.val = ck.atd(localIdx);
+            sparseItems.add(item);
+            nzCount++;
+        }
+        return nzCount;
+    }
+
+    /**
+     * convert a set of H2O chunks (representing a part of a vector) to a sparse DMatrix
+     * @param response name of the response column
+     * @param weight name of the weight column
+     * @param fold name of the fold assignment column
+     * @return DMatrix
+     * @throws XGBoostError
+     */
+    static DMatrix convertChunksToDMatrix(Key<DataInfo> dataInfoKey,
+                                          Chunk[] chunks,
+                                          int response,
+                                          int weight,
+                                          int fold,
+                                          boolean sparse) throws XGBoostError {
+        long nRows = chunks[0]._len;
+
+        DMatrix trainMat;
+
+        DataInfo di = dataInfoKey.get();
+
+        float[] resp = new float[(int)nRows];
+        float[] weights = null;
+        if(-1 != weight) {
+            weights = new float[(int)nRows];
+        }
+        try {
+            if (sparse) {
+                Log.info("Treating matrix as sparse.");
+                // 1 0 2 0
+                // 4 0 0 3
+                // 3 1 2 0
+                boolean csc = false; //di._cats == 0;
+
+                // truly sparse matrix - no categoricals
+                // collect all nonzeros column by column (in parallel), then stitch together into final data structures
+                if (csc) {
+                    trainMat = csc(chunks, weight, nRows, di, resp, weights);
+                } else {
+                    trainMat = csr(chunks, weight, response, (int) nRows, di, resp, weights);
+                }
+            } else {
+                trainMat = dense(chunks, weight, di, response, resp, weights);
+            }
+        } catch (NegativeArraySizeException e) {
+            throw new IllegalArgumentException(technote(11, "Data is too large to fit into the 32-bit Java float[] array that needs to be passed to the XGBoost C++ backend. Use H2O GBM instead."));
+        }
+
+        int len = (int) trainMat.rowNum();
+        resp = Arrays.copyOf(resp, len);
+        trainMat.setLabel(resp);
+        if (weight!=-1){
+            weights = Arrays.copyOf(weights, len);
+            trainMat.setWeight(weights);
+        }
+//    trainMat.setGroup(null); //fold //FIXME - only needed if CV is internally done in XGBoost
```

Comment:
Just an idea - maybe we could create JIRAs right away for todos so these todo's don't get lost in the code